engine/schema/src/com/cloud/dc/DataCenterVO.java

```diff
@@ -365,6 +366,18 @@ public void setLocalStorageEnabled(boolean enabled) {
     }
 
     @Override
+    public boolean isDynamicRoutingEnabled() {
+        String dynr = getDetail(OSPFZoneConfig.Params.ENABLED.name());
```

Comment:
There is a common pattern across this patch where ``getDetail`` is called and then the returned value is checked.  If the return value is ``null``, a default value is returned.  Please consider adding a ``getOptionalDetail(key, defaultValue)`` method that encapsulates this logic.  If this method is added, please add a unit test to verify its behavior.